### PR TITLE
Fix order of premises in integer induction

### PR DESCRIPTION
**Pull request by mezpusz**
The main change in this PR is that for bounded induction a certain order of premises was expected during binary resolution with the clausified induction formula so that binary resolution can find the right instantiated literals to resolve:

1. the main premise is binary resolved with substitution `{ x -> t }` where `x` is the induction variable and `t` is the induction term
2. the bounds are resolved with empty substitution since 1. instantiates `x` to `t` in the rest of clause

This order was not maintained and hence most bounded inductions were not properly resolved, possibly delaying the necessary binary resolution forever.

The second contribution is unit testing for the current induction framework (including tests not related to the bug):

1. Bounded induction has multiple premises from indices, so we cherry picked parts of unit testing code from branch `inequality-resolution` to allow for adding clauses to indices.
2. Induction creates possibly many clauses, so the permutation matching algorithm matching the generated clauses to expected clauses was improved to work on many clauses without timeouting.

### CODE REVIEW

File: Inferences/Induction.cpp
Comment (MichaelRawson) [reply Inferences/Induction.cpp:713]: Looking closer: could this be stack-allocated like `static RobSubstitution subst;`? Looks like it's only used in this function, and all dangling references are cleaned up. If so we can avoid an allocation/indirection here.

**Review by MichaelRawson — APPROVED** [review]
OK, looks good, only a few nitpicks. I'm assuming this does what you think it does - it looks plausible. :-) I also haven't looked in detail at the testing code, I will defer to Johannes to have a look at that, but I expect he will be pleased.

File: Inferences/Induction.cpp
Comment (mezpusz) [reply Inferences/Induction.cpp:699]: Apparently this doesn't work not because the type cannot be deduced from the second argument but rather due to `List` not begin a typename in itself without the template parameters. But surprisingly (to me), `toResolve->push(..., toResolve)` works. Is that good enough or should we leave it?

Comment (MichaelRawson) [reply Inferences/Induction.cpp:699]: Hmm, maybe even just `push` (no qualifiers) would work because of Koenig lookup. Just a thought, if not I think the original is most clear.

Comment (mezpusz) [reply Inferences/Induction.cpp:699]: Yes, that would be the cleanest. My compiler doesn't find the function this way unfortunately. I guess this is partly why "static" functions in STL like `make_pair`, `make_unique` and so on are defined outside the template classes...

Comment (MichaelRawson) [reply Inferences/Induction.cpp:699]: Boo. Never mind, thanks for checking!

File: Test/TestUtils.hpp
Comment (joe-hauns) [inline Test/TestUtils.hpp:257]: This function is unused, right?

**Review by joe-hauns — COMMENTED** [review]
Had a look at the testing code, and commented the one thing I'm not sure about. Let me know what u think!

File: Test/TestUtils.hpp
Comment (mezpusz) [reply Test/TestUtils.hpp:257]: Yes, good catch! I'll remove it.

File: Test/GenerationTester.hpp
Comment (mezpusz) [reply Test/GenerationTester.hpp:105]: I guess you mean we could instead pass all necessary options through ctors into `Induction`. It can be done, although it would be tedious to use due to the increasing number of options for induction. Maybe separating all induction options into a separate object inside `Options` would be the nicest way and we could pass this single object through ctors.

I don't know if it aligns with the effort to make the current options more hierarchical though. Also, it would increase the size of change in this PR significantly. What do you think?

Comment (joe-hauns) [reply Test/GenerationTester.hpp:105]: Yess i kind of thought that might be the case. 
If it's quite some work, don't mind. 
I think we'd kind of need a bigger refactor to untangle the options code from other parts of the code base..

Comment (mezpusz) [reply Test/GenerationTester.hpp:105]: I ended up using `getOptions()` in `Induction` from `InferenceEngine` which doesn't use the global options. But I couldn't eliminate modifying `env.options` from `GenerationTester` due to a massive amount of static calls in `InductionHelper` which is used by term indices as well, and would ultimately require supplying the test indices themselves with the options from the test... At this point the modifications where so huge that I guess a separate PR would be much cleaner.

Comment (joe-hauns) [reply Test/GenerationTester.hpp:105]: Fair enough! Looks all neat and good to go!

**Review by joe-hauns — APPROVED** [review]

File: Inferences/Induction.cpp
Comment (MichaelRawson) [reply Inferences/Induction.cpp]: Hey @mezpusz - any news on this before it gets merged?

Comment (mezpusz) [reply Inferences/Induction.cpp]: Sorry, I just changed it to your suggestion without replying. It should be OK, I tested it and there were no issues.

**Pull request merged**
